src/opdcodes.rs

```diff
@@ -771,4 +929,89 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn test_store() -> Result<(), VMError> {
+        let mut vm = setup_vm();
+
+        // Set up value in source register (R1)
+        let value_to_store = 0x4242;
+        vm.write_register(1, value_to_store);
+
+        // Calculate target address (PC + offset)
+        let pc_offset = 2;
+        let target_address = vm.registers.pc.wrapping_add(pc_offset);
+
+        // Create ST instruction: ST R1, #2
+        // Format: 0011 001 000000010
+        // 0011 = ST opcode
+        // 001 = source register (R1)
+        // 000000010 = PC offset of 2
+        let instruction = 0b0011_001_000000010;
+
+        store(&mut vm, instruction)?;
+
+        // Verify value was stored in memory at target address
+        assert_eq!(vm.read_memory(target_address)?, value_to_store);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_store_indirect() -> Result<(), VMError> {
+        let mut vm = setup_vm();
+
+        // Set up value in source register (R1)
+        let value_to_store = 0x4242;
+        vm.write_register(1, value_to_store);
+
+        // Set up pointer in memory
+        let pointer_offset = 2;
+        let pointer_addr = vm.registers.pc.wrapping_add(pointer_offset);
+        let final_addr = 0x3100;
+        vm.write_memory(pointer_addr, final_addr)?;
+
+        // Create STI instruction: STI R1, #2
+        // Format: 1011 001 000000010
+        // 1011 = STI opcode
+        // 001 = source register (R1)
+        // 000000010 = PC offset of 2
+        let instruction = 0b1011_001_000000010;
+
+        store_indirect(&mut vm, instruction)?;
+
+        // Verify value was stored in memory at final address
+        assert_eq!(vm.read_memory(final_addr)?, value_to_store);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_store_register() -> Result<(), VMError> {
+        let mut vm = setup_vm();
+
+        // Set up base register (R1) with base address
+        let base_address = 0x3000;
+        vm.write_register(1, base_address);
+
+        // Set up value in source register (R2)
+        let value_to_store = 0x4242;
+        vm.write_register(2, value_to_store);
+
+        // Create STR instruction: STR R2, R1, #2
+        // Format: 0111 010 001 000010
+        // 0111 = STR opcode
+        // 010 = source register (R2)
+        // 001 = base register (R1)
+        // 000010 = offset of 2
+        let instruction = 0b0111_010_001_000010;
+
+        store_register(&mut vm, instruction)?;
+
+        // Calculate target address and verify value was stored
+        let target_address = base_address.wrapping_add(2);
+        assert_eq!(vm.read_memory(target_address)?, value_to_store);
+
+        Ok(())
+    }
```

Comment:
Lovely tests! I like how you wrote a small description explaining what instruction it will compute.

Maybe, in a future PR, you could add tests that test operands working together rather than having a unit test for every operand.

Reply:
Thanks Fabri, next pr contains test checking that different operands works together